README.md
# Office Addin Sideloader

[![PyPI](https://img.shields.io/pypi/v/oaloader?style=flat-square)](https://pypi.org/project/oaloader/)
![GitHub](https://img.shields.io/github/license/elonzh/office-addin-sideloader?style=flat-square)

A handy tool to manage your office addins locally,
you can use it for addin development or deploy your addins for your clients out of AppSource.

> NOTE: currently only support windows.

## Features

- Add or remove Office Add-in locally.
- Support local or url manifest source.
- Debug sideload status and list manifest info.
- Single binary without any dependency.
- Use it as a library.
- Generate add-in installer/uninstaller with [sentry](https://sentry.io) support by single command.
- Support fixing add-in [APP ERROR](https://docs.microsoft.com/en-us/office365/troubleshoot/installation/cannot-install-office-add-in) and [clearing cache](https://docs.microsoft.com/en-us/office/dev/add-ins/testing/clear-cache).

## Installation

### Pre-built releases

If you just use the command line and don't have a python environment,
download pre-built binary from [GitHub Releases](https://github.com/elonzh/office-addin-sideloader/releases).

### Pypi

```shell
> pip install oaloader
```

## Quick Start

```text
> ./oaloader.exe --help
Usage:  [OPTIONS] COMMAND [ARGS]...

  Manage your office addins locally.

Options:
  --version         Show the version and exit.
  -l, --level TEXT  The log level  [default: info]
  --help            Show this message and exit.

Commands:
  add     Register catalog and add manifests, manifests can be file paths
          or...

  fix     Try fixing `APP ERROR` when starting up add-ins.
  info    Debug sideload status.
  remove  Remove manifest from catalog and manifest can be a file path or...
```

## Build an Addin installer

1. Install [Poetry](https://python-poetry.org/docs/).
2. Run `poetry install` to prepare environment.
3. Checkout [Nuitka Requirements](https://nuitka.net/doc/user-manual.html#requirements) and install a C compiler.
4. Run `poetry run invoke installer -m <YOUR-ADDIN-MANIFEST-URL>` to build your own installer.

If your want customize the installer, just edit `installer.jinja2` or write your own installer with `oaloader` module.

## Build an Addin uninstaller

Just using invoke `uninstaller` task like `installer` above.

## FAQ

### How it works?

https://docs.microsoft.com/en-us/office/dev/add-ins/testing/create-a-network-shared-folder-catalog-for-task-pane-and-content-add-ins

### Get error like `ImportError: DLL load failed while importing win32xxx` when import pywin32 module.

Try this solution:

1. Open a terminal as Administrator
2. Get your virtualenv path by running `poetry env info`
3. Run `poetry run python <virtualenv path>/.venv/Scripts/pywin32_postinstall.py -install`

see https://github.com/mhammond/pywin32/issues/1431#issuecomment-548584385 for more details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.4] - 2021-10-22
### Fixed
- Office subkey may not exist.

## [0.4.3] - 2021-10-13
### Fixed
- Fix potential ssl issue when using urlopen.

## [0.4.2] - 2021-09-08
### Changed
- Exception in `fix_app_error` and `clear_cache` will be captured silently.
### Fixed
- Fix setting sentry breadcrumb data wrongly.

## [0.4.1] - 2021-08-26
### Changed
- `setup_sentry` now `set_tag` about office installation and use default integrations.
- installer/uninstaller shows version now.
### Fixed
- Fix `fix_app_error` deletes wrong keys.

## [0.4.0] - 2021-08-24
### Added
- Support fixing add-in and clear cache.
- Support build uninstaller.

### Fixed
- Fix office_installation error.
- Update Nuitka to develop version to fix sign issue.
- more robust installer/uninstaller behavior.

## [0.3.0] - 2021-08-12
### Added
- Support check office installation.
- Add dry run mode for installer task.

## [0.2.1] - 2021-08-12
### Changed
- Return unknown when find office installation failed.

## [0.2.0] - 2021-08-09
### Added
- Add sentry_sdk support for installer.
- Support debug system info and office installation.

## [0.1.0] - 2021-08-05
### Added
- Release to pypi.


tasks.py
import hashlib
import re
import sys
from pathlib import Path

import jinja2
from invoke import Context, Result, task

nuitka_base_args = [
    sys.executable,
    "-m",
    "nuitka",
    "--assume-yes-for-downloads",
    "--onefile",
    "--output-dir=build",
    "--remove-output",
    # Windows Defender will report Trojan:Win32/Sabsik.TE.A!ml
    # when using --nofollow-imports
    # "--nofollow-imports",
]


def rename_executable(output: str, executable: str):
    matched = re.search(r"Successfully created '(.+)'\.", output)
    if not matched:
        return
    src = Path(matched.group(1))
    dst = src.with_name(executable + src.suffix)
    dst.unlink(missing_ok=True)
    dst = src.rename(dst)
    print(f"Rename executable to {dst}")
    return dst


@task()
def build(c, executable="oaloader"):
    """
    Build oaloader.
    """
    c: Context

    args = nuitka_base_args.copy()
    args.extend(["oaloader/__main__.py"])
    r: Result = c.run(" ".join(args))
    rename_executable(r.stdout, executable)


def make_executable(
    c,
    template,
    manifests,
    success_msg,
    fail_msg,
    sentry_dsn="",
    executable="addin-executable",
    icon="",
    dry=False,
):
    c: Context
    env = jinja2.Environment()
    env.filters["repr"] = repr
    t = env.from_string(Path(template).read_text())
    script = t.render(
        manifests=manifests,
        success_msg=success_msg,
        fail_msg=fail_msg,
        sentry_dsn=sentry_dsn,
    )

    p = Path(f"executable_{hashlib.md5(script.encode()).hexdigest()[:8]}").with_suffix(
        ".py"
    )
    p.write_text(script, encoding="utf-8")
    args = nuitka_base_args.copy()
    args.extend(
        [
            "--include-package=sentry_sdk",
            "--windows-uac-admin",
            # "--windows-disable-console",
        ]
    )
    if icon:
        args.extend(
            [
                "--windows-icon-from-ico",
                icon,
            ]
        )
    args.append(str(p))
    r: Result = c.run(" ".join(args), dry=dry)
    if dry:
        print(f"Dry run mode, delete file {p!s} manually")
        return executable
    else:
        e = rename_executable(r.stdout, executable)
        p.unlink(missing_ok=True)
        return e


@task(iterable=["manifests"])
def installer(
    c,
    manifests,
    success_msg="Addins are successfully installed. Press any key to continue...",
    fail_msg="Operation failed, do you have a valid Microsoft installation?",
    sentry_dsn="",
    executable="addin-installer",
    icon="",
    template="bin/installer.jinja2",
    dry=False,
):
    """
    Build an addin installer.
    """
    return make_executable(
        c,
        template=template,
        manifests=manifests,
        success_msg=success_msg,
        fail_msg=fail_msg,
        sentry_dsn=sentry_dsn,
        executable=executable,
        icon=icon,
        dry=dry,
    )


@task(iterable=["manifests"])
def uninstaller(
    c,
    manifests,
    success_msg="Addins are successfully uninstalled. Press any key to continue...",
    fail_msg="Operation failed, do you have a valid Microsoft installation?",
    sentry_dsn="",
    executable="addin-uninstaller",
    icon="",
    template="bin/uninstaller.jinja2",
    dry=False,
):
    """
    Build an addin uninstaller.
    """
    return make_executable(
        c,
        template=template,
        manifests=manifests,
        success_msg=success_msg,
        fail_msg=fail_msg,
        sentry_dsn=sentry_dsn,
        executable=executable,
        icon=icon,
        dry=dry,
    )


oaloader/core.py
import os
import platform
import re
import shutil
import winreg
from pathlib import Path

import pywintypes
import requests
import win32com.client
import win32net
import win32netcon
import xmltodict
from loguru import logger

from . import const

__all__ = [
    "local_server_url",
    "get_net_shares",
    "add_net_share",
    "remove_net_share",
    "find_catalog",
    "add_catalog",
    "remove_catalog",
    "open_office_sub_key",
    "enum_reg",
    "load_manifest",
    "add_manifests",
    "remove_manifests",
    "system_info",
    "office_installation",
    "fix_app_error",
    "clear_cache",
]


def local_server_url(netname):
    local_server = win32net.NetServerGetInfo(None, 100)
    url = rf"\\{local_server['name'].lower()}\{netname}"
    return url


def get_net_shares(share_type=win32netcon.STYPE_DISKTREE):
    net_shares, _, _ = win32net.NetShareEnum(None, 2)
    if share_type is not None:
        net_shares = list(filter(lambda v: v["type"] == share_type, net_shares))
    return net_shares


def add_net_share(netname: str, path) -> str:
    """
    https://docs.microsoft.com/en-us/windows/win32/api/lmshare/nf-lmshare-netshareadd
    """
    path = Path(path).absolute()
    if not path.exists():
        path.mkdir(parents=True, exist_ok=True)

    path_str = str(path)
    share_info = {
        "netname": netname,
        "path": path_str,
        "type": win32netcon.STYPE_DISKTREE,
        "permissions": win32netcon.ACCESS_READ,
        "remark": "",
        "max_uses": -1,
        "current_uses": 0,
        "passwd": "",
    }
    url = local_server_url(netname)
    context_logger = logger.bind(netname=netname, url=url, path=path_str)
    try:
        win32net.NetShareAdd(None, 2, share_info)
        context_logger.debug("add net share")
    except win32net.error as e:
        #  (2118, 'NetShareAdd', '名称已经共享。')
        if e.winerror != 2118:
            raise
        context_logger.debug("net share already exists")
    return url


def remove_net_share(netname: str):
    rv = win32net.NetShareDel(None, netname)
    logger.bind(netname=netname).debug("remove net share")
    return rv


def find_catalog(root, url: str):
    for item in enum_reg(root):
        if item["attribute"] == "Url" and item["value"] == url:
            _id = item["key"]
            return _id
    logger.bind(url=url).debug("can not find catalog")


def add_catalog(root, url: str, hide: bool = False):
    _id = find_catalog(root, url) or str(pywintypes.CreateGuid())
    key = winreg.CreateKey(root, _id)
    winreg.SetValueEx(key, "Id", 0, winreg.REG_SZ, _id)
    winreg.SetValueEx(key, "Url", 0, winreg.REG_SZ, url)
    # show in menu
    winreg.SetValueEx(
        key,
        "Flags",
        0,
        winreg.REG_DWORD,
        int(not hide),
    )
    logger.bind(id=_id, url=url, hide=hide).debug("add catalog")
    return key


def remove_catalog(root, url: str):
    _id = find_catalog(root, url)
    if _id:
        winreg.DeleteKey(root, _id)
        logger.bind(key=_id, url=url).debug("remove catalog")


def open_office_sub_key(sub_key):
    # open office key first make sure the office installation exists
    office_key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, const.SUBKEY_OFFICE)
    catalog_key = winreg.CreateKey(office_key, sub_key)
    winreg.CloseKey(office_key)
    return catalog_key


def enum_reg(root):
    keys = []
    i = 0
    while True:
        try:
            keys.append(winreg.EnumKey(root, i))
            i += 1
        except winreg.error as e:
            # [WinError 259] 没有可用的数据了。
            if e.winerror == 259:
                break
            else:
                raise

    for key in keys:
        k = winreg.OpenKey(root, key)
        i = 0
        while True:
            try:
                a, v, t = winreg.EnumValue(k, i)
                yield {
                    "key": key,
                    "attribute": a,
                    "value": v,
                    "type": t,
                }
                i += 1
            except winreg.error as e:
                # [WinError 259] 没有可用的数据了。
                if e.winerror == 259:
                    break
                else:
                    raise


url_pattern = re.compile(
    r"https?://(www.)?[-a-zA-Z0-9@:%._+~#=]{1,256}.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()!@:%_+.~#?&/=]*)"
)


def is_url(s: str) -> bool:
    return bool(url_pattern.match(s))


def load_manifest(src: str):
    if is_url(src):
        logger.bind(url=src).debug("load manifest from url")
        raw = requests.get(src).content.decode(const.MANIFEST_ENCODING)
    else:
        raw = Path(src).read_text(encoding=const.MANIFEST_ENCODING)
    data = xmltodict.parse(raw)["OfficeApp"]
    return {
        "id": data["Id"],  # must have
        "version": data.get("Version"),
        "provider_name": data.get("ProviderName"),
        "display_name": data.get("DisplayName", {}).get("@DefaultValue"),
        "description": data.get("Description", {}).get("@DefaultValue"),
    }, raw


def add_manifests(
    netname=const.DEFAULT_NETNAME,
    path=const.DEFAULT_PATH,
    hide=False,
    manifests=tuple(),
):
    """
    Register catalog and add manifests, manifests can be file paths or urls.

    NOTE: run as admin.
    """
    root = open_office_sub_key(const.OFFICE_SUBKEY_CATALOG)
    url = add_net_share(netname, path)
    add_catalog(root, url, hide)

    for m in manifests:
        d, raw = load_manifest(m)
        dst = path.joinpath(f"{d['id']}.xml")
        dst.write_text(raw, encoding=const.MANIFEST_ENCODING)
        logger.bind(src=m, dst=str(dst)).debug("add manifest")


def remove_manifests(
    netname=const.DEFAULT_NETNAME, rm_all=False, rm_catalog=False, manifests=tuple()
):
    """
    Remove manifest from catalog and manifest can be a file path or url.

    NOTE: run as admin.
    """
    try:
        share = win32net.NetShareGetInfo(None, netname, 2)
    except win32net.error as e:
        # 2310, 'NetShareGetInfo', '共享资源不存在。'
        if e.winerror == 2310:
            logger.bind(netname=netname).info("net share does not exist")
            return
        raise
    path = Path(share["path"])

    if rm_all:
        manifests = path.glob("*.xml")
    for m in manifests:
        d, _ = load_manifest(str(m))
        dst = path.joinpath(f"{d['id']}.xml")
        dst.unlink(missing_ok=True)
        logger.bind(src=m, dst=str(dst)).debug("remove manifest")

    if rm_catalog:
        root = open_office_sub_key(const.OFFICE_SUBKEY_CATALOG)
        url = local_server_url(netname)
        remove_catalog(root, url)
        remove_net_share(netname)


def fix_app_error():
    """
    https://docs.microsoft.com/en-us/office365/troubleshoot/installation/cannot-install-office-add-in
    """
    root = open_office_sub_key(const.OFFICE_SUBKEY_PROVIDER)
    for v in enum_reg(root):
        if v["attribute"] == "UniqueId" and (
            v["value"] == "Anonymous" or v["value"].endswith("_ADAL")
        ):
            k = v["key"]
            logger.bind(subykey=k, attribute=v["attribute"], value=v["value"]).debug(
                "delete invalid provider key"
            )
            winreg.DeleteKey(root, k)


def clear_cache():
    """
    https://docs.microsoft.com/en-us/office/dev/add-ins/testing/clear-cache
    """
    paths = [
        Path(os.getenv("LOCALAPPDATA")).joinpath(r"Microsoft\Office\16.0\Wef"),
        Path(os.getenv("USERPROFILE")).joinpath(
            r"AppData\Local\Packages\Microsoft.Win32WebViewHost_cw5n1h2txyewy\AC\#!123\INetCache"
        ),
    ]

    def onerror(func, path, exc_info):
        log = logger.bind(func=func, path=path, exc_info=exc_info)
        level = "WARNING"
        if exc_info[0] == FileNotFoundError:
            level = "DEBUG"
        log.log(level, "rmtree failed")

    for p in paths:
        logger.bind(path=str(p)).debug("rmtree")
        shutil.rmtree(
            p,
            onerror=onerror,
        )


def system_info():
    return dict(
        platform=platform.platform(),
        system=platform.system(),
        release=platform.release(),
        version=platform.version(),
    )


def office_installation(app: str):
    try:
        word = win32com.client.Dispatch(f"{app.capitalize()}.Application")
        # some attrs may not exist
        return dict(
            name=getattr(word, "Name", "Unknown"),
            version=getattr(word, "Version", "Unknown"),
            build=getattr(word, "Build", "Unknown"),
            path=getattr(word, "Path", "Unknown"),
            startup_path=getattr(word, "StartupPath", "Unknown"),
        )
    except pywintypes.com_error as e:
        # com_error: (-2147221005, '无效的类字符串', None, None)
        # com_error: (-2146959355, '服务器运行失败', None, None)
        if e.hresult in (-2147221005, -2146959355):
            return dict(
                name="Unknown",
                version="Unknown",
                build="Unknown",
                path="Unknown",
                startup_path="Unknown",
            )
        raise


oaloader/command.py
from pathlib import Path

import click
from tabulate import tabulate

from .const import (
    APP_NAME,
    DEFAULT_NETNAME,
    DEFAULT_PATH,
    OFFICE_SUBKEY_CATALOG,
    OFFICE_SUBKEY_PROVIDER,
    SUBKEY_OFFICE,
    version,
)
from .core import (
    add_manifests,
    clear_cache,
    enum_reg,
    fix_app_error,
    get_net_shares,
    load_manifest,
    local_server_url,
    office_installation,
    open_office_sub_key,
    remove_manifests,
    system_info,
)
from .log import setup_logger

__all__ = ["cli"]

opt_netname = click.option(
    "--netname",
    "-n",
    type=click.STRING,
    default=DEFAULT_NETNAME,
    show_default=True,
    help="NETNAME represents a share folder or a office add-in catalog.",
)
opt_path = click.option(
    "--path",
    "-p",
    type=click.Path(file_okay=False),
    default=DEFAULT_PATH,
    show_default=True,
    help="The directory stores manifests and is shared as NETNAME.",
)
arg_manifests = click.argument(
    "manifests",
    type=click.STRING,
    nargs=-1,
)


def use_decorators(*decorators):
    def wrapper(function):
        for d in reversed(decorators):
            function = d(function)
        return function

    return wrapper


@click.group()
@click.version_option(version=version, prog_name=APP_NAME)
@click.option(
    "--level",
    "-l",
    type=click.STRING,
    default="info",
    show_default=True,
    help="The log level",
)
def cli(level: str):
    """
    Manage your office add-ins locally.
    """
    setup_logger(level)


@cli.command()
@use_decorators(opt_netname, opt_path)
@click.option("--hide", is_flag=True, help="Hide add-ins in the catalog.")
@use_decorators(arg_manifests)
def add(netname, path, hide, manifests):
    """
    Register catalog and add manifests, manifests can be file paths or urls.

    NOTE: run as admin.
    """
    add_manifests(netname, path, hide, manifests)
    clear_cache()


@cli.command()
@use_decorators(opt_netname)
@click.option(
    "--all", "-a", "rm_all", is_flag=True, help="Remove all manifests in catalog path."
)
@click.option("--catalog", "-c", "rm_catalog", is_flag=True, help="Remove the catalog.")
@use_decorators(arg_manifests)
def remove(netname, rm_all, rm_catalog, manifests):
    """
    Remove manifest from catalog and manifest can be a file path or url.

    NOTE: run as admin.
    """
    remove_manifests(netname, rm_all, rm_catalog, manifests)
    clear_cache()


@cli.command()
def fix():
    """
    Try fixing `APP ERROR` when starting up add-ins.
    """
    fix_app_error()
    clear_cache()


def _echo_table(title, table_data):
    click.secho(title, bold=True, fg="green")
    if table_data:
        click.echo(tabulate(table_data, headers="keys"))
    else:
        click.secho("Nothing", fg="yellow")
    click.echo()


def _safe_open_key(title, sub_key):
    try:
        return open_office_sub_key(sub_key)
    except FileNotFoundError:
        click.secho(title, fg="green")
        raise click.ClickException(
            click.style(
                "Can't find Office 16 registry key, do you have a valid Microsoft installation?",
                fg="yellow",
            )
        )


@cli.command()
@use_decorators(opt_path)
def info(path):
    """
    Debug sideload status.
    """

    _echo_table("System Info:", [system_info()])
    _echo_table("Word Installation:", [office_installation("word")])

    net_shares = get_net_shares()
    _echo_table("Net Shares:", net_shares)

    title = rf"HKEY_CURRENT_USER\{SUBKEY_OFFICE}\{OFFICE_SUBKEY_PROVIDER}:"
    rv = filter(
        lambda v: v["attribute"] == "UniqueId",
        enum_reg(_safe_open_key(title, OFFICE_SUBKEY_PROVIDER)),
    )

    _echo_table(title, rv)

    title = rf"HKEY_CURRENT_USER\{SUBKEY_OFFICE}\{OFFICE_SUBKEY_CATALOG}:"
    rv = enum_reg(_safe_open_key(title, OFFICE_SUBKEY_CATALOG))
    _echo_table(title, rv)

    urls = []
    for item in rv:
        if item["attribute"] == "Url":
            urls.append(item["value"])

    url_to_path = {local_server_url(v["netname"]): v["path"] for v in net_shares}
    paths = set()
    paths.add(Path(path))
    for u in urls:
        p = url_to_path.get(u)
        if p:
            paths.add(Path(p))

    for path in paths:
        addins = []
        for p in path.glob("*.xml"):
            d, _ = load_manifest(str(p))
            addins.append(d)
        _echo_table(f"Office Add-ins in `{path}`:", addins)


oaloader.py
import os
import shutil
import sys
import winreg
from pathlib import Path

import click
import pywintypes
import win32net
import win32netcon
import xmltodict
from loguru import logger
from tabulate import tabulate

server = None  # Run on local machine.
subkey = r"Software\Microsoft\Office\16.0\WEF\TrustedCatalogs"

__version__ = "0.0.1"
app_name = "office-addin-sideloader"
app_dir = Path(click.get_app_dir(app_name))

default_path = app_dir.joinpath("addins")
default_netname = "office-addins"


def setup_logger(level: str):
    level = level.upper()
    logger.remove()
    fmt = (
        "<green>{time:YYYY-MM-DD HH:mm:ss.SSS}</green> | "
        "<level>{level: <8}</level> | "
        "<level>{message}</level> - "
        "{extra}"
    )
    logger.add(
        sys.stdout,
        level=level,
        format=fmt,
    )
    logger.add(
        app_dir.joinpath("main.log"),
        level="DEBUG",
        format=fmt,
    )


def local_server_url(netname):
    local_server = win32net.NetServerGetInfo(None, 100)
    url = rf"\\{local_server['name'].lower()}\{netname}"
    return url


def get_net_shares(share_type=win32netcon.STYPE_DISKTREE):
    net_shares, _, _ = win32net.NetShareEnum(server, 2)
    if share_type is not None:
        net_shares = list(filter(lambda v: v["type"] == share_type, net_shares))
    return net_shares


def add_net_share(netname: str, path) -> str:
    """
    https://docs.microsoft.com/en-us/windows/win32/api/lmshare/nf-lmshare-netshareadd
    """
    path = os.path.abspath(path)
    share_info = {
        "netname": netname,
        "path": path,
        "type": win32netcon.STYPE_DISKTREE,
        "permissions": win32netcon.ACCESS_READ,
        "remark": "",
        "max_uses": -1,
        "current_uses": 0,
        "passwd": "",
    }
    url = local_server_url(netname)
    context_logger = logger.bind(netname=netname, url=url, path=path)
    try:
        win32net.NetShareAdd(server, 2, share_info)
        context_logger.debug("add net share")
    except win32net.error as e:
        #  (2118, 'NetShareAdd', '名称已经共享。')
        if e.winerror != 2118:
            raise
        context_logger.debug("net share already exists")
    return url


def remove_net_share(netname: str):
    rv = win32net.NetShareDel(server, netname)
    logger.bind(netname=netname).debug("remove net share")
    return rv


def find_catalog(root, url: str):
    for item in enum_reg(root):
        if item["attribute"] == "Url" and item["value"] == url:
            _id = item["key"]
            return _id
    logger.bind(url=url).debug("can not find catalog")


def add_catalog(root, url: str, hide: bool = False):
    _id = find_catalog(root, url) or str(pywintypes.CreateGuid())
    key = winreg.CreateKey(root, _id)
    winreg.SetValueEx(key, "Id", 0, winreg.REG_SZ, _id)
    winreg.SetValueEx(key, "Url", 0, winreg.REG_SZ, url)
    # show in menu
    winreg.SetValueEx(
        key,
        "Flags",
        0,
        winreg.REG_DWORD,
        int(not hide),
    )
    logger.bind(id=_id, url=url, hide=hide).debug("add catalog")
    return key


def remove_catalog(root, url: str):
    _id = find_catalog(root, url)
    if _id:
        winreg.DeleteKey(root, _id)
        logger.bind(key=_id, url=url).debug("remove catalog")


def enum_reg(root):
    keys = []
    i = 0
    while True:
        try:
            keys.append(winreg.EnumKey(root, i))
            i += 1
        except winreg.error as e:
            # [WinError 259] 没有可用的数据了。
            if e.winerror == 259:
                break
            else:
                raise

    rv = []
    for key in keys:
        k = winreg.OpenKey(root, key)
        i = 0
        while True:
            try:
                a, v, t = winreg.EnumValue(k, i)
                rv.append(
                    {
                        "key": key,
                        "attribute": a,
                        "value": v,
                        "type": t,
                    }
                )
                i += 1
            except winreg.error as e:
                # [WinError 259] 没有可用的数据了。
                if e.winerror == 259:
                    break
                else:
                    raise
    return rv


def load_manifest(path):
    path = Path(path)
    manifest = xmltodict.parse(path.read_text())["OfficeApp"]
    return {
        "file": path.name,
        "id": manifest["Id"],  # must have
        "version": manifest.get("Version"),
        "provider_name": manifest.get("ProviderName"),
        "display_name": manifest.get("DisplayName", {}).get("@DefaultValue"),
        "description": manifest.get("Description", {}).get("@DefaultValue"),
    }


opt_netname = click.option(
    "--netname", "-n", type=click.STRING, default=default_netname
)
opt_path = click.option(
    "--path", "-p", type=click.Path(file_okay=False), default=default_path
)


def use_options(*options):
    def wrapper(function):
        for option in reversed(options):
            function = option(function)
        return function

    return wrapper


@click.group()
@click.version_option(version=__version__, prog_name=app_name)
@click.option("--level", "-l", type=click.STRING, default="info", help="log level")
def cli(level: str):
    """
    manage your office addins locally.
    """
    setup_logger(level)


@cli.command()
@use_options(opt_netname, opt_path)
@click.option("--hide", is_flag=True, help="hide addins in the catalog")
@click.argument(
    "manifests", type=click.Path(exists=True, file_okay=True, dir_okay=False), nargs=-1
)
def add(netname, path, hide, manifests):
    """
    add manifest to catalog.

    NOTE: run as admin.
    """
    path = Path(path)
    if not path.exists():
        path.mkdir(parents=True, exist_ok=True)
    url = add_net_share(netname, path)
    root = winreg.OpenKey(winreg.HKEY_CURRENT_USER, subkey)
    add_catalog(root, url, hide)

    for m in manifests:
        d = load_manifest(m)
        dst = path.joinpath(f"{d['id']}.xml")
        shutil.copy(m, dst)
        logger.bind(src=m, dst=str(dst)).debug("copy manifest")


@cli.command()
@use_options(opt_netname)
@click.option("--all", "-a", "rm_all", is_flag=True, help="remove all manifests")
@click.option("--catalog", "-c", "rm_catalog", is_flag=True, help="remove the catalog")
@click.argument(
    "manifests", type=click.Path(exists=True, file_okay=True, dir_okay=False), nargs=-1
)
def remove(netname, rm_all, rm_catalog, manifests):
    """
    remove manifest from catalog.

    NOTE: run as admin.
    """
    try:
        share = win32net.NetShareGetInfo(server, netname, 2)
    except win32net.error as e:
        # 2310, 'NetShareGetInfo', '共享资源不存在。'
        if e.winerror == 2310:
            logger.info("net share does not exist")
            return
        raise
    path = Path(share["path"])

    if rm_all:
        manifests = path.glob("*.xml")
    for m in manifests:
        d = load_manifest(m)
        dst = path.joinpath(f"{d['id']}.xml")
        dst.unlink(missing_ok=True)
        logger.bind(src=m, dst=str(dst)).debug("remove manifest")

    if rm_catalog:
        root = winreg.OpenKey(winreg.HKEY_CURRENT_USER, subkey)
        url = local_server_url(netname)
        remove_catalog(root, url)
        remove_net_share(netname)


def _echo_table(title, table_data):
    click.secho(title, bold=True, fg="green")
    if table_data:
        click.echo(tabulate(table_data, headers="keys"))
    else:
        click.secho("Nothing", fg="yellow")
    click.echo()


@cli.command()
@use_options(opt_path)
def info(path):
    """
    debug sideload status.
    """
    root = winreg.OpenKey(winreg.HKEY_CURRENT_USER, subkey)
    rv = enum_reg(root)
    _echo_table(rf"HKEY_CURRENT_USER\{subkey}:", rv)

    urls = []
    for item in rv:
        if item["attribute"] == "Url":
            urls.append(item["value"])

    net_shares = get_net_shares()

    _echo_table("Net Shares:", net_shares)

    url_to_path = {local_server_url(v["netname"]): v["path"] for v in net_shares}
    paths = set()
    paths.add(Path(path))
    for u in urls:
        p = url_to_path.get(u)
        if p:
            paths.add(Path(p))

    for path in paths:
        addins = []
        for p in path.glob("*.xml"):
            addins.append(load_manifest(p))
        _echo_table(f"Office Addins in `{path}`:", addins)


if __name__ == "__main__":
    cli()
